utils/fixed_decimal/src/decimal.rs

```diff
@@ -746,6 +756,238 @@ impl FixedDecimal {
+    /// # use std::str::FromStr;
+    ///
+    /// let mut dec = FixedDecimal::from_str("-1.5").unwrap();
+    /// dec.half_truncate_right(0);
+    /// assert_eq!("-1", dec.to_string());
+    ///
+    /// let mut dec = FixedDecimal::from_str("0.4").unwrap();
+    /// dec.half_truncate_right(0);
+    /// assert_eq!("0", dec.to_string());
+    ///
+    /// let mut dec = FixedDecimal::from_str("0.5").unwrap();
+    /// dec.half_truncate_right(0);
+    /// assert_eq!("0", dec.to_string());
+    ///
+    /// let mut dec = FixedDecimal::from_str("0.6").unwrap();
+    /// dec.half_truncate_right(0);
+    /// assert_eq!("1", dec.to_string());
+    ///    
+    /// let mut dec = FixedDecimal::from_str("1.5").unwrap();
+    /// dec.half_truncate_right(0);
+    /// assert_eq!("1", dec.to_string());
+    ///
+    /// let mut dec = FixedDecimal::from_str("3.954").unwrap();
+    /// dec.half_truncate_right(0);
+    /// assert_eq!("4", dec.to_string());
+    /// ```
+    pub fn half_truncate_right(&mut self, position: i16) {
+        let digit_after_n = self.digit_at_next_positon(position);
+        let condition = {
+            if digit_after_n < 5 {
+                false
+            } else if digit_after_n > 5 {
+                true
+            } else if position == i16::MIN {
+                false
+            } else if self.nonzero_magnitude_right() < position - 1 {
+                true
+            } else {
+                false
+            }
+        };
+
+        self.expand_with_condition(position, condition);
+    }
+
+    /// This performs the normat expand.
+    /// However, if the user set the condition to `false`, the expand will not increment the digit by 1.
+    /// This is useful for implementing methods such as half_expand and half_trunc.
+    pub fn expand_with_condition(&mut self, position: i16, condition: bool) {
+        let before_truncate_is_zero = self.is_zero();
+        let before_truncate_is_bottom_magnitude = self.nonzero_magnitude_right();
+        let before_truncate_magnitude = self.magnitude;
+        self.truncate_right(position);
+
+        if before_truncate_is_zero || position <= before_truncate_is_bottom_magnitude || !condition
+        {
+            #[cfg(debug_assertions)]
+            self.check_invariants();
+            return;
+        }
+
+        if position <= before_truncate_magnitude {
+            let result = self.increment_abs_by_one();
+            if result.is_err() {
+                // Do nothing for now.
+            }
+
+            #[cfg(debug_assertions)]
+            self.check_invariants();
+            return;
+        }
+
+        debug_assert!(self.digits.is_empty());
+        self.digits.push(1);
+        self.magnitude = position;
+        self.upper_magnitude = cmp::max(self.upper_magnitude, position);
+
+        #[cfg(debug_assertions)]
+        self.check_invariants();
+    }
+
+    /// Take the expand of the number at a particular position.
+    ///
+    /// # Examples
```

Comment:
Nit: Docs-test examples should be more concise. If there are more than approximately 10 to 15 lines, you should probably make a unit test instead.

You can file a follow-up if you prefer.

Reply:
I moved all the big test cases to a test block at the end of the page. I will use a .json file for the tests in the next PR.